### ISSUE TITLE
False-positive for UseDataClass: `copy` shouldn't be considered a default function

### ISSUE THREAD

**Issue by mickele** [arturbosch/detekt#1040]
One of the rules to check if we should use a data class is to check if the class only contains "default functions": https://github.com/arturbosch/detekt/blob/master/detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UseDataClass.kt#L56

I understand why `hashCode`, `equals` and `toString` are there, but Kotlin's data class `copy` is not an overridable function.

False-positive class:
```
class Clipboard(private val type: Int) {
    fun copy(text: String) = Unit
}
```

I think `copy` could be removed from the list of default functions since it's not overridable.

What do you think?

**Comment by schalkms** [comment]
Please note that there's a reason for that behavior. Detekt adheres to the offical Kotlin guideline [here](https://kotlinlang.org/docs/reference/data-classes.html).
A data class automatically generates a `copy()` method. If your class `Clipboard` has a custom `copy()` method, `copy()` of the **data class** `Clipboard` will differ from the original implementation. Hence, there's a check for that in the rule `UseDataClass`.

**Comment by schalkms** [comment]
It's not related to the `override` modifier at all.

**Comment by marianosimone** [comment]
I think @mickele refers to a `copy` method that it has logic related to the job of a `Clipboard`, and not to what `copy` means in the `data class` world in Kotlin.

**Comment by schalkms** [comment]
@marianosimone I hadn't thought of that. However, I think that methods shouldn't have the same name as reserved ones, which is the case for `copy()`.

**Comment by mickele** [comment]
Yep, I understood that it's not related to `override` itself, I just meant those 3 methods are present in all objects and usually overriding them is not a good idea, hence `UseDataClass`. But a method copy is not present in any class by default, it's a nice benefit of using data classes but in this example I'm not even returning the object itself, `copy` is just a normal function of a `Clipboard` class.

Is `copy` a reserved name? From their docs they only prohibits if the signature is the same, which is not in this case.

**Comment by schalkms** [comment]
No, Kotlin doesn't prohibit one from implementing a `copy()` method, which is not related to copying an object at all. It wouldn't be that easy to detect whether an `copy()` method is related to copying an object or not. One thing which comes to my mind is that the rule can check if the copy() method returns an object of type `Clipboard`. This check would work for some cases but not for all.

**Comment by schalkms** [comment]
My approach requires type solving though, which detekt doesn't have atm.
